fms/models/roberta.py

```diff
@@ -0,0 +1,321 @@
+
+
+@dataclass
+class RoBERTaConfig(ModelConfig):
+    src_vocab_size: int = 50265
+    emb_dim: int = 768
+    nheads: int = 12
+    nlayers: int = 12
+    pad_id: int = 1
+    hidden_grow_factor: float = 4.0
+    activation_fn: str = "gelu"
+    classifier_activation_fn: str = "tanh"
+    max_pos: int = 512
+    p_dropout: float = 0.1
+    multiquery_attn: bool = False
+    norm_eps: float = 1e-12
+    tie_heads: bool = False
+
+
+class RoBERTaBlock(nn.Module):
+    def __init__(self, config: RoBERTaConfig):
+        super().__init__()
+        self.config = config
+
+        self.ln = nn.LayerNorm(self.config.emb_dim, self.config.norm_eps)
+        self.ff_ln = nn.LayerNorm(self.config.emb_dim, self.config.norm_eps)
+
+        self.attn = MultiHeadAttention(
+            self.config.emb_dim,
+            self.config.emb_dim // self.config.nheads,
+            self.config.emb_dim // self.config.nheads,
+            self.config.nheads,
+            kvheads=1 if self.config.multiquery_attn else self.config.nheads,
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        self.ff_sub_layer = FeedForwardBlock(
+            self.config.emb_dim,
+            hidden_grow_factor=self.config.hidden_grow_factor,
+            activation_fn=str_to_activation(self.config.activation_fn),
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        *,
+        mask: Optional[torch.Tensor] = None,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # first we do MHA
+        residual = x
+        # self attention
+        x = self.attn(
+            q=x,
+            k=x,
+            v=x,
+            mask=mask,
+            attn_algorithm=attn_algorithm,
+            is_self=True,
+            is_causal_mask=False,
+        )
+
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+
+        # residual connection
+        x = x + residual
+        # post ln
+        x = self.ln(x)
+
+        # then we do FF and Add&Norm
+        residual = x
+        x = self.ff_sub_layer(x)
+
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+
+        # another residual
+        x = x + residual
+        x = self.ff_ln(x)
+
+        return x
+
+
+class RoBERTaHeadless(nn.Module):
+    def __init__(
+        self, config: RoBERTaConfig, distributed_strategy: DistributedStrategy
+    ):
+        super().__init__()
+        self.config = config
+
+        self.layers = nn.ModuleList(
+            [
+                distributed_strategy.distribute_layer(RoBERTaBlock(self.config), i)
+                for i in range(self.config.nlayers)
+            ]
+        )
+
+        self.embedding = nn.Embedding(self.config.src_vocab_size, self.config.emb_dim)
```

Comment:
hmm... in the llama model we do TP-wrap nn.Embedding modules (via TPWordEmbedding) with `self.emb = nn.Embedding(self.vocab_size, self.emb_dim // world_size)` and code like `apply_embedding_tp(self.emb, we.emb, self.world_size, self.rank)`